appengine/pond_storage/opponents.py

```diff
@@ -19,27 +19,115 @@
 
 import cgi
 import json
+import math
+import random
 from google.appengine.ext import ndb
 from pond_storage import *
 
-def get_opponent_ducks(duck):
-  """Returns list of opponent ducks."""
-  # TODO: Add filter for published ducks
-  # TODO: Add logic here to pick the duck based on the ranking
-  user_key = get_user_key(users.get_current_user())
-  duck_query = Duck.query()
-  duck_list = []
-  for duck in duck_query:
-    if duck.key.parent() != user_key:
-      duck_list.append(get_duck_info(duck))
-  return duck_list[0:3]
+START_PERCENT = .10
+END_PERCENT = .15
+NEARBY_PERCENT = .10
+DUCKS_NEEDED = 3
+
+def get_entries_in_range(user_duck, entries, start_rank, end_rank):
+  """Get all entries in the given range that"""
+  user_entry = user_duck.leaderboard_entry_key.get()
+  entryQuery = LeaderboardEntry.query()
+  entryQuery = entryQuery.filter(LeaderboardEntry.ranking >= start_rank)
+  entryQuery = entryQuery.filter(LeaderboardEntry.ranking <= end_rank)
+  entryQuery = entryQuery.filter(LeaderboardEntry.ranking != user_entry.ranking)
+
+  for entry in entries:
+    entryQuery = entryQuery.filter(LeaderboardEntry.ranking != entry.ranking)
+  newEntries = entryQuery.fetch()
+  return newEntries
+
+def get_nearby_entries(user_duck, cur_rank, entries, total_entries, num_needed):
+  """Get ducks with similar rankings to the users duck"""
+  nearby_entries = []
+  percent_range = NEARBY_PERCENT
+
+  while(len(nearby_entries) < num_needed and percent_range <= 1):
+    min_rank = int(math.floor(cur_rank - total_entries * percent_range))
+    max_rank = int(math.ceil(cur_rank + total_entries * percent_range))
+    nearby_entries = get_entries_in_range(user_duck, entries, min_rank, max_rank)
+    percent_range += .20
+  return random.sample(nearby_entries, num_needed)
+
+def get_worse_ducks(user_duck, entries, cur_rank, total_entries):
```

Comment:
get_worse_ducks and get_better_ducks both return entries not ducks, so I think they should be named like get_nearby_entries